README.md
![Java CI](https://github.com/BioDWH2/BioDWH2-ArcadeDB-Server/workflows/Java%20CI/badge.svg?branch=develop) ![Release](https://img.shields.io/github/v/release/BioDWH2/BioDWH2-ArcadeDB-Server) ![Downloads](https://img.shields.io/github/downloads/BioDWH2/BioDWH2-ArcadeDB-Server/total) ![License](https://img.shields.io/github/license/BioDWH2/BioDWH2-ArcadeDB-Server)

# BioDWH2-ArcadeDB-Server
**BioDWH2** is an easy-to-use, automated, graph-based data warehouse and mapping tool for bioinformatics and medical informatics. The main repository can be found [here](https://github.com/BioDWH2/BioDWH2).

This repository contains the **BioDWH2-ArcadeDB-Server** utility which can be used to create and explore an ArcadeDB graph database from any BioDWH2 workspace. There is no need for any ArcadeDB installation. All necessary components are bundled with this tool.

## Download
The latest release version of **BioDWH2-ArcadeDB-Server** can be downloaded [here](https://github.com/BioDWH2/BioDWH2-ArcadeDB-Server/releases/latest).

## Usage
BioDWH2-ArcadeDB-Server requires the Java Runtime Environment 11+. The JRE 11 archive is available [here](https://www.oracle.com/java/technologies/javase/jdk11-archive-downloads.html).

Creating a database from any workspace is done using the following command. Every time the workspace is updated or changed, the create command has to be executed again.
~~~BASH
> java -jar BioDWH2-ArcadeDB-Server.jar --create /path/to/workspace
~~~

Once the database has been created, the database and ArcadeDB Studio can be started as follows:
~~~BASH
> java -jar BioDWH2-ArcadeDB-Server.jar --start /path/to/workspace
~~~

By default, two users are created for ArcadeDB Studio:
 * Username: `root`, Password: `biodwh2-arcadedb`
 * Username: `biodwh2`, Password: `biodwh2`

## Help
~~~
Usage: BioDWH2-ArcadeDB-Server.jar [-h] [-c <workspacePath>] [-cs <workspacePath>]
                                   [-s <workspacePath>] [-p <port>] [-sp <studioPort>]
  -h, --help                            print this message
  -s, --start <workspacePath>           Start an ArcadeDB server for the workspace
  -c, --create <workspacePath>          Create an ArcadeDB database from the workspace graph
  -cs, --create-start <workspacePath>   Create and start an ArcadeDB database from the workspace graph
  -p, --port <port>                     Specifies the ArcadeDB server port(-range) (default 2480-2489)
~~~

CHANGELOG.md
# Changelog

## 📦 Version [v1.0.1](https://github.com/BioDWH2/BioDWH2-ArcadeDB-Server/releases/tag/v1.0.1)

Released: 13.09.2022

* ```[feature]``` Improve logging
* ```[feature]``` Add additional default user
* ```[fix]``` Fix edge creation
* ```[chore]``` Update dependencies

## 📦 Version [v1.0.0](https://github.com/BioDWH2/BioDWH2-ArcadeDB-Server/releases/tag/v1.0.0)

Released: 01.06.2022

 * ```[feature]``` Initial release


src/src/main/java/de/unibi/agbi/biodwh2/arcadedb/server/HashUtils.java
package de.unibi.agbi.biodwh2.arcadedb.server;

import org.apache.commons.codec.digest.DigestUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.attribute.BasicFileAttributes;

final class HashUtils {
    private HashUtils() {
    }

    static String getMd5HashFromFile(final String filePath) throws IOException {
        try (final InputStream inputStream = Files.newInputStream(Paths.get(filePath))) {
            return DigestUtils.md5Hex(inputStream);
        }
    }

    static String getFastPseudoHashFromFile(final String filePath) throws IOException {
        final BasicFileAttributes attributes = Files.readAttributes(Paths.get(filePath), BasicFileAttributes.class);
        return DigestUtils.md5Hex(attributes.lastModifiedTime() + "__" + attributes.size());
    }
}


src/src/main/java/de/unibi/agbi/biodwh2/arcadedb/server/ArcadeDBServer.java
package de.unibi.agbi.biodwh2.arcadedb.server;

import de.unibi.agbi.biodwh2.core.net.BioDWH2Updater;
import de.unibi.agbi.biodwh2.arcadedb.server.model.CmdArgs;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import picocli.CommandLine;

import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class ArcadeDBServer {
    private static final Logger LOGGER = LoggerFactory.getLogger(ArcadeDBServer.class);

    private ArcadeDBServer() {
    }

    public static void main(final String... args) {
        final CmdArgs commandLine = parseCommandLine(args);
        new ArcadeDBServer().run(commandLine);
    }

    private static CmdArgs parseCommandLine(final String... args) {
        final CmdArgs result = new CmdArgs();
        final CommandLine cmd = new CommandLine(result);
        cmd.parseArgs(args);
        return result;
    }

    private void run(final CmdArgs commandLine) {
        BioDWH2Updater.checkForUpdate("BioDWH2-ArcadeDB-Server",
                                      "https://api.github.com/repos/BioDWH2/BioDWH2-ArcadeDB-Server/releases");
        if (commandLine.createStart != null)
            createAndStartWorkspaceServer(commandLine);
        else if (commandLine.start != null)
            startWorkspaceServer(commandLine);
        else if (commandLine.create != null)
            createWorkspaceDatabase(commandLine);
        else
            printHelp(commandLine);
    }

    private void createAndStartWorkspaceServer(final CmdArgs commandLine) {
        final String workspacePath = commandLine.createStart;
        if (!verifyWorkspaceExists(workspacePath)) {
            printHelp(commandLine);
            return;
        }
        final ArcadeDBService service = new ArcadeDBService(workspacePath);
        service.deleteOldDatabase();
        service.startArcadeDBService(commandLine.port);
        service.createDatabase();
        storeWorkspaceHash(workspacePath);
        service.openBrowser();
    }

    private boolean verifyWorkspaceExists(final String workspacePath) {
        if (StringUtils.isEmpty(workspacePath) || !Paths.get(workspacePath).toFile().exists()) {
            if (LOGGER.isErrorEnabled())
                LOGGER.error("Workspace path '" + workspacePath + "' was not found");
            return false;
        }
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Using workspace directory '" + workspacePath + "'");
        return true;
    }

    private void printHelp(final CmdArgs commandLine) {
        CommandLine.usage(commandLine, System.out);
    }

    private void storeWorkspaceHash(final String workspacePath) {
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Updating workspace ArcadeDB cache checksum...");
        final Path hashFilePath = Paths.get(workspacePath, "arcadedb/checksum.txt");
        try {
            final String hash = HashUtils.getFastPseudoHashFromFile(
                    Paths.get(workspacePath, "sources/mapped.db").toString());
            final FileWriter writer = new FileWriter(hashFilePath.toFile());
            writer.write(hash);
            writer.close();
        } catch (IOException e) {
            if (LOGGER.isErrorEnabled())
                LOGGER.error("Failed to store hash of workspace mapped graph", e);
        }
    }

    private void startWorkspaceServer(final CmdArgs commandLine) {
        final String workspacePath = commandLine.start;
        if (!verifyWorkspaceExists(workspacePath)) {
            printHelp(commandLine);
            return;
        }
        if (!checkArcadeDBDatabaseMatchesWorkspace(workspacePath) && LOGGER.isInfoEnabled())
            LOGGER.warn("The ArcadeDB database is out-of-date and should be recreated with the --create command");
        final ArcadeDBService service = new ArcadeDBService(workspacePath);
        service.startArcadeDBService(commandLine.port);
        service.openBrowser();
    }

    private boolean checkArcadeDBDatabaseMatchesWorkspace(final String workspacePath) {
        try {
            final String hash = HashUtils.getFastPseudoHashFromFile(
                    Paths.get(workspacePath, "sources/mapped.db").toString());
            final Path hashFilePath = Paths.get(workspacePath, "arcadedb/checksum.txt");
            if (Files.exists(hashFilePath)) {
                final String storedHash = new String(Files.readAllBytes(hashFilePath)).trim();
                return hash.equals(storedHash);
            }
        } catch (IOException e) {
            if (LOGGER.isWarnEnabled())
                LOGGER.warn("Failed to check hash of workspace mapped graph", e);
        }
        return false;
    }

    private void createWorkspaceDatabase(final CmdArgs commandLine) {
        final String workspacePath = commandLine.create;
        if (!verifyWorkspaceExists(workspacePath)) {
            printHelp(commandLine);
            return;
        }
        final ArcadeDBService service = new ArcadeDBService(workspacePath);
        service.deleteOldDatabase();
        service.startArcadeDBService(commandLine.port);
        service.createDatabase();
        storeWorkspaceHash(workspacePath);
        service.stopArcadeDBService();
    }
}


src/src/main/java/de/unibi/agbi/biodwh2/arcadedb/server/ArcadeDBService.java
package de.unibi.agbi.biodwh2.arcadedb.server;

import com.arcadedb.ContextConfiguration;
import com.arcadedb.GlobalConfiguration;
import com.arcadedb.database.DatabaseInternal;
import com.arcadedb.database.RID;
import com.arcadedb.graph.MutableEdge;
import com.arcadedb.graph.MutableVertex;
import com.arcadedb.graph.Vertex;
import com.arcadedb.index.IndexException;
import com.arcadedb.index.lsm.LSMTreeIndexAbstract;
import com.arcadedb.schema.*;
import com.arcadedb.server.ArcadeDBServer;
import de.unibi.agbi.biodwh2.core.model.graph.Edge;
import de.unibi.agbi.biodwh2.core.model.graph.Graph;
import de.unibi.agbi.biodwh2.core.model.graph.IndexDescription;
import de.unibi.agbi.biodwh2.core.model.graph.Node;
import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.StringUtils;
import org.json.JSONArray;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.awt.*;
import java.io.IOException;
import java.lang.reflect.Array;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

/**
 * <a href="https://docs.arcadedb.com/#Embed-Server">ArcadeDB Embedded Server</a>
 */
public class ArcadeDBService {
    private static final Logger LOGGER = LoggerFactory.getLogger(ArcadeDBService.class);

    private final String workspacePath;
    private final Path databasePath;
    private ArcadeDBServer server;

    public ArcadeDBService(final String workspacePath) {
        this.workspacePath = workspacePath;
        databasePath = Paths.get(workspacePath, "arcadedb");
        InjectionLogger.inject();
    }

    public void startArcadeDBService(String port) {
        port = validatePort(port);
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Starting ArcadeDB DBMS on localhost:" + port + "...");
        server = new ArcadeDBServer(getServerConfig(port));
        server.start();
        createBioDWH2UserIfNotExists();
    }

    private String validatePort(String port) {
        final String fallback = "2480-2489";
        port = port.trim().replace(" ", "");
        if (port.contains("-")) {
            final String[] rangeParts = StringUtils.split(port, '-');
            if (rangeParts.length == 2) {
                return port;
            } else
                LOGGER.warn("Failed to parse port or port range '" + port + "', falling back to '" + fallback + "'");
        }
        try {
            final int portNumber = Integer.parseInt(port);
            return String.valueOf(Math.abs(portNumber));
        } catch (NumberFormatException ignored) {
            LOGGER.warn("Failed to parse port or port range '" + port + "', falling back to '" + fallback + "'");
        }
        return fallback;
    }

    private ContextConfiguration getServerConfig(final String port) {
        final ContextConfiguration config = new ContextConfiguration();
        config.setValue(GlobalConfiguration.HA_SERVER_LIST, "localhost");
        config.setValue(GlobalConfiguration.HA_REPLICATION_INCOMING_HOST, "0.0.0.0");
        config.setValue(GlobalConfiguration.HA_ENABLED, false);
        config.setValue(GlobalConfiguration.SERVER_NAME, "BioDWH2-ArcadeDB-Server");
        config.setValue(GlobalConfiguration.SERVER_HTTP_INCOMING_PORT, port);
        config.setValue(GlobalConfiguration.SERVER_ROOT_PASSWORD, "biodwh2-arcadedb");
        config.setValue(GlobalConfiguration.SERVER_ROOT_PATH, databasePath);
        return config;
    }

    private void createBioDWH2UserIfNotExists() {
        if (!server.getSecurity().existsUser("biodwh2")) {
            final JSONObject user = new JSONObject();
            user.put("name", "biodwh2");
            user.put("password", server.getSecurity().encodePassword("biodwh2"));
            final JSONObject databases = new JSONObject();
            final JSONArray databasePrivilege = new JSONArray();
            databasePrivilege.put("admin");
            databases.put("*", databasePrivilege);
            user.put("databases", databases);
            server.getSecurity().createUser(user);
        }
    }

    public void stopArcadeDBService() {
        if (server != null)
            server.stop();
    }

    public void deleteOldDatabase() {
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Removing old database...");

        try {
            FileUtils.deleteDirectory(databasePath.toFile());
        } catch (IOException e) {
            if (LOGGER.isErrorEnabled())
                LOGGER.error("Failed to remove old database '" + databasePath + "'", e);
        }
    }

    public void createDatabase() {
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Creating ArcadeDB database...");
        final DatabaseInternal db = server.createDatabase("BioDWH2");
        db.getConfiguration().setValue(GlobalConfiguration.BUCKET_DEFAULT_PAGE_SIZE, 1048576);
        try (Graph graph = new Graph(Paths.get(workspacePath, "sources/mapped.db"), true)) {
            Files.createDirectories(databasePath);
            final HashMap<Long, RID> nodeIdArcadeDBIdMap = new HashMap<>();
            createNodes(db, graph, nodeIdArcadeDBIdMap);
            createEdges(db, graph, nodeIdArcadeDBIdMap);
            createIndices(db, graph);
        } catch (IOException e) {
            if (LOGGER.isErrorEnabled())
                LOGGER.error("Failed to create ArcadeDB database '" + databasePath + "'", e);
        }
    }

    private void createNodes(final DatabaseInternal db, final Graph graph,
                             final HashMap<Long, RID> nodeIdArcadeDBIdMap) {
        final String[] labels = graph.getNodeLabels();
        for (int i = 0; i < labels.length; i++) {
            final String label = labels[i];
            if (LOGGER.isInfoEnabled())
                LOGGER.info("Creating nodes with label '" + label + "' (" + (i + 1) + "/" + labels.length + ")...");
            // Create a node definition for the label
            final VertexType definition = db.getSchema().createVertexType(label);
            final Map<String, de.unibi.agbi.biodwh2.core.lang.Type> propertyKeyTypes = graph.getPropertyKeyTypesForNodeLabel(
                    label);
            for (final String key : propertyKeyTypes.keySet())
                if (!Node.IGNORED_FIELDS.contains(key))
                    definition.createProperty(key, getTypeByPropertyType(propertyKeyTypes.get(key)));
            // Create the actual nodes
            for (final Node node : graph.getNodes(label)) {
                final MutableVertex arcadeNode = db.newVertex(label).modify();
                for (final String propertyKey : node.keySet())
                    setPropertySafe(node, arcadeNode, propertyKey);
                final RID id = arcadeNode.save().getIdentity();
                nodeIdArcadeDBIdMap.put(node.getId(), id);
            }
        }
    }

    private Type getTypeByPropertyType(de.unibi.agbi.biodwh2.core.lang.Type propertyType) {
        Type type = Type.getTypeByClass(propertyType.getType());
        if (type == null && propertyType.isList()) {
            Class<?> componentType = propertyType.getComponentType();
            // Fallback to string if component type is unknown
            if (componentType == null)
                componentType = String.class;
            type = Type.getTypeByClass(Array.newInstance(componentType, 0).getClass());
        }
        return type;
    }

    private void setPropertySafe(final Node node, final MutableVertex arcadeNode, final String propertyKey) {
        try {
            if (!Node.IGNORED_FIELDS.contains(propertyKey)) {
                Object value = node.getProperty(propertyKey);
                if (value instanceof Collection)
                    value = convertCollectionToArray((Collection<?>) value);
                if (value != null)
                    arcadeNode.set(propertyKey, value);
            }
        } catch (Exception e) {
            if (LOGGER.isWarnEnabled())
                LOGGER.warn(
                        "Illegal property '" + propertyKey + " -> " + node.getProperty(propertyKey) + "' for node '" +
                        node.getId() + "[:" + node.getLabel() + "]'");
        }
    }

    @SuppressWarnings({"SuspiciousToArrayCall"})
    private Object convertCollectionToArray(final Collection<?> collection) {
        Class<?> type = null;
        for (Object t : collection) {
            if (t != null) {
                type = t.getClass();
                break;
            }
        }
        if (type != null) {
            if (type.equals(String.class))
                return collection.stream().map(type::cast).toArray(String[]::new);
            if (type.equals(Boolean.class))
                return collection.stream().map(type::cast).toArray(Boolean[]::new);
            if (type.equals(Integer.class))
                return collection.stream().map(type::cast).toArray(Integer[]::new);
            if (type.equals(Float.class))
                return collection.stream().map(type::cast).toArray(Float[]::new);
            if (type.equals(Long.class))
                return collection.stream().map(type::cast).toArray(Long[]::new);
            if (type.equals(Double.class))
                return collection.stream().map(type::cast).toArray(Double[]::new);
            if (type.equals(Byte.class))
                return collection.stream().map(type::cast).toArray(Byte[]::new);
            if (type.equals(Short.class))
                return collection.stream().map(type::cast).toArray(Short[]::new);
        }
        return collection.stream().map(Object::toString).toArray(String[]::new);
    }

    private void createEdges(final DatabaseInternal db, final Graph graph,
                             final HashMap<Long, RID> nodeIdArcadeDBIdMap) {
        final String[] labels = graph.getEdgeLabels();
        for (int i = 0; i < labels.length; i++) {
            final String label = labels[i];
            if (LOGGER.isInfoEnabled())
                LOGGER.info("Creating edges with label '" + label + "' (" + (i + 1) + "/" + labels.length + ")...");
            // Create an edge definition for the label
            final EdgeType definition = db.getSchema().createEdgeType(label);
            final Map<String, de.unibi.agbi.biodwh2.core.lang.Type> propertyKeyTypes = graph.getPropertyKeyTypesForEdgeLabel(
                    label);
            for (final String key : propertyKeyTypes.keySet())
                if (!Edge.IGNORED_FIELDS.contains(key))
                    definition.createProperty(key, getTypeByPropertyType(propertyKeyTypes.get(key)));
            // Create the actual edges
            for (final Edge edge : graph.getEdges(label)) {
                final Vertex fromNode = db.lookupByRID(nodeIdArcadeDBIdMap.get(edge.getFromId()), false).asVertex(
                        false);
                final RID toNodeId = nodeIdArcadeDBIdMap.get(edge.getToId());
                final MutableEdge arcadeEdge = fromNode.newEdge(edge.getLabel(), toNodeId, false).modify();
                for (final String propertyKey : edge.keySet())
                    if (!Edge.IGNORED_FIELDS.contains(propertyKey)) {
                        Object value = edge.getProperty(propertyKey);
                        if (value instanceof Collection)
                            value = convertCollectionToArray((Collection<?>) value);
                        if (value != null)
                            arcadeEdge.set(propertyKey, value);
                    }
                arcadeEdge.save();
            }
        }
    }

    private void createIndices(final DatabaseInternal db, final Graph graph) {
        if (LOGGER.isInfoEnabled())
            LOGGER.info("Creating indices...");
        final IndexDescription[] indices = graph.indexDescriptions();
        for (final IndexDescription index : indices) {
            if (index.isArrayProperty()) {
                if (LOGGER.isInfoEnabled())
                    LOGGER.info("Skipping " + index.getType() + " index creation on '" + index.getProperty() +
                                "' field for " + index.getTarget() + " label '" + index.getLabel() +
                                "' as array indices are not yet supported by ArcadeDB");
                continue;
            }
            if (LOGGER.isInfoEnabled())
                LOGGER.info("Creating " + index.getType() + " index on '" + index.getProperty() + "' field for " +
                            index.getTarget() + " label '" + index.getLabel() + "'...");
            final boolean isUnique = index.getType() == IndexDescription.Type.UNIQUE;
            try {
                db.getSchema().createTypeIndex(Schema.INDEX_TYPE.LSM_TREE, isUnique, index.getLabel(),
                                               new String[]{index.getProperty()}, LSMTreeIndexAbstract.DEF_PAGE_SIZE,
                                               LSMTreeIndexAbstract.NULL_STRATEGY.SKIP, null);
            } catch (IndexException e) {
                LOGGER.warn("Error during " + index.getType() + " index creation on '" + index.getProperty() +
                            "' field for " + index.getTarget() + " label '" + index.getLabel() + "'");
            }
        }
    }

    public void openBrowser() {
        final int port = server.getHttpServer().getPort();
        try {
            Desktop.getDesktop().browse(new URI("http://localhost:" + port + "/"));
        } catch (IOException | URISyntaxException e) {
            if (LOGGER.isErrorEnabled())
                LOGGER.error("Failed to open Browser", e);
        }
    }
}


src/src/main/java/de/unibi/agbi/biodwh2/arcadedb/server/InjectionLogger.java
package de.unibi.agbi.biodwh2.arcadedb.server;

import com.arcadedb.log.LogManager;
import com.arcadedb.log.Logger;
import org.slf4j.LoggerFactory;

import java.util.logging.Level;

class InjectionLogger implements Logger {
    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(ArcadeDBService.class);

    public static void inject() {
        LogManager.instance().setLogger(new InjectionLogger());
    }

    @Override
    public void log(Object requester, Level level, String message, Throwable exception, String context,
                    Object arg1, Object arg2, Object arg3, Object arg4, Object arg5, Object arg6, Object arg7,
                    Object arg8, Object arg9, Object arg10, Object arg11, Object arg12, Object arg13,
                    Object arg14, Object arg15, Object arg16, Object arg17) {
        final String requesterName;
        if (requester instanceof String)
            requesterName = (String) requester;
        else if (requester instanceof Class<?>)
            requesterName = ((Class<?>) requester).getName();
        else if (requester != null)
            requesterName = requester.getClass().getName();
        else
            requesterName = "com.arcadedb";
        message = '[' + requesterName + "] " + String.format(message, arg1, arg2, arg3, arg4, arg5, arg6, arg7,
                                                             arg8, arg9, arg10, arg11, arg12, arg13, arg14,
                                                             arg15, arg16, arg17);
        if (level == Level.INFO)
            LOGGER.info(message, exception);
        else if (level == Level.WARNING)
            LOGGER.warn(message, exception);
        else if (level == Level.SEVERE)
            LOGGER.error(message, exception);
        else if (level == Level.FINE)
            LOGGER.debug(message, exception);
    }

    @Override
    public void log(Object requester, Level level, String message, Throwable exception, String context,
                    Object... args) {
        final String requesterName;
        if (requester instanceof String)
            requesterName = (String) requester;
        else if (requester instanceof Class<?>)
            requesterName = ((Class<?>) requester).getName();
        else if (requester != null)
            requesterName = requester.getClass().getName();
        else
            requesterName = "com.arcadedb";
        message = '[' + requesterName + "] " + String.format(message, args);
        if (level == Level.INFO)
            LOGGER.info(message, exception);
        else if (level == Level.WARNING)
            LOGGER.warn(message, exception);
        else if (level == Level.SEVERE)
            LOGGER.error(message, exception);
        else if (level == Level.FINE)
            LOGGER.debug(message, exception);
    }

    @Override
    public void flush() {
    }
}


src/src/main/java/de/unibi/agbi/biodwh2/arcadedb/server/model/CmdArgs.java
package de.unibi.agbi.biodwh2.arcadedb.server.model;

import picocli.CommandLine;

@CommandLine.Command(name = "BioDWH2-ArcadeDB-Server.jar", sortOptions = false, separator = " ", footer = "Visit https://biodwh2.github.io for more documentation.")
public class CmdArgs {
    @CommandLine.Option(names = {"-h", "--help"}, usageHelp = true, description = "print this message", order = 1)
    public boolean help;
    @CommandLine.Option(names = {
            "-s", "--start"
    }, arity = "1", paramLabel = "<workspacePath>", description = "Start an ArcadeDB server for the workspace", order = 2)
    public String start;
    @CommandLine.Option(names = {
            "-c", "--create"
    }, arity = "1", paramLabel = "<workspacePath>", description = "Create an ArcadeDB database from the workspace graph", order = 3)
    public String create;
    @CommandLine.Option(names = {
            "-cs", "--create-start"
    }, arity = "1", paramLabel = "<workspacePath>", description = "Create and start an ArcadeDB database from the workspace graph", order = 4)
    public String createStart;
    @CommandLine.Option(names = {
            "-p", "--port"
    }, defaultValue = "2480-2489", paramLabel = "<port>", description = "Specifies the ArcadeDB server port(-range) (default 2480-2489)", order = 5)
    public String port;
}
